xmppserver/src/main/java/org/jivesoftware/openfire/group/GroupManager.java

```diff
@@ -789,6 +625,423 @@ private void evictCachedUserSharedGroups() {
+        if (wasMember) {
+            params.put("member", admin.toString());
+            GroupEventDispatcher.dispatchEvent(updatedGroup, GroupEventDispatcher.EventType.member_removed, params);
+        }
+        GroupEventDispatcher.dispatchEvent(updatedGroup, GroupEventDispatcher.EventType.admin_added, params);
+    }
+
+    /**
+     * Updates the caches maintained by this manager and dispatches events that reflect an admin user having been
+     * removed from a group.
+     *
+     * This method is intended to be invoked to update the internal state of GroupManager after a group has been
+     * updated. It should rarely be invoked by code that is not part of the GroupManager implementation.
+     *
+     * @param group The group that was modified.
+     * @param admin The address of the admin that was removed from the group.
+     */
+    public void adminRemovedPostProcess(@Nonnull final Group group, @Nonnull final JID admin)
+    {
+        evictCachedUserForGroup(admin);
+        final Group updatedGroup;
+        try {
+            // By forcing the lookup, the cache will automatically receive the required refresh.
+            updatedGroup = getGroup(group.getName(), true);
+        } catch (GroupNotFoundException e) {
+            Log.error("Group '{}' was not found after an admin was removed from it. This is indicative of a bug in Openfire. Please consider reporting it.", group.getName(), e);
+            return;
+        }
+
+        // Fire event.
+        final Map<String, String> params = new HashMap<>();
+        params.put("admin", admin.toString());
+
+        GroupEventDispatcher.dispatchEvent(updatedGroup, GroupEventDispatcher.EventType.admin_removed, params);
+    }
+
+    /**
+     * Updates the caches maintained by this manager and dispatches events that reflect a member user having been added
+     * to a group.
+     *
+     * This method is intended to be invoked to update the internal state of GroupManager after a group has been
+     * updated. It should rarely be invoked by code that is not part of the GroupManager implementation.
+     *
+     * @param group The group that was modified.
+     * @param member The address of the member that was added to the group.
+     * @param wasAdmin Whether the member was an admin of the group (which are mutually exclusive roles) before the update.
+     */
+    public void memberAddedPostProcess(@Nonnull final Group group, @Nonnull final JID member, final boolean wasAdmin)
+    {
+        evictCachedUserForGroup(member);
+        final Group updatedGroup;
+        try {
+            // By forcing the lookup, the cache will automatically receive the required refresh.
+            updatedGroup = getGroup(group.getName(), true);
+        } catch (GroupNotFoundException e) {
+            Log.error("Group '{}' was not found after a member was added to it. This is indicative of a bug in Openfire. Please consider reporting it.", group.getName(), e);
```

Comment:
This might be also just a co-incidence, if someone deletes the group at about the same time someone is added to it.

Is it possible to single thread some of these actions? Adding a member requires a read lock, deleting the group a read-write lock? Suspect it's tricky across the cluster.

Probably more effort than it's worth.

Reply:
I considered that, but given the complexity of the code, and the way how groups are nested, possibly retrieved from external sources (AD/LDAP), and used all over the place, and clustering, as you mentioned, my fear of introducing deadlocks won the argument.